### DISCUSSION TITLE
upgrade to 14 and Listeners

### DISCUSSION THREAD

**Discussion by macit-emre** [orchidsoftware/platform#2619] [Q&A] [answered]
My Listeners work on version 13 cant be upgraded to 14 because of the change usages of Listeners. 
İ have selectbox fields on my page like "Choose Type" 
Then when you change your type into option 1, i am showing 1 input and 1 selectbox more. 
if you choose option 2, another flow, additionally i am showing only 1 selectbox more.
Cant make it work on orchid 14 alpha7

**I Want**

  1. Old style listener with multiple fields depending on other fields.

Any eligible coding option to resolve my issue is appreciated.

**Additional context**
I want to upgrade to orchid 14 but this is the only issue for my upgrade process

**Comment by tabuna** [comment]
Hi @macit-emre. The variability of the new Listener record is still preserved.
Could you share an example of what is no longer working?

**Reply by macit-emre** [reply]
i need to change a part of the form layout based on the XHR request. I am currently doing it now. But when i switch to alpha7, only handle method works. layout method not triggering and changes the layout anymore. i need to change the form element or add another one, based on the dropdown value selected in the form. 


![Screenshot 2023-05-30 at 23 23 04](https://github.com/orchidsoftware/platform/assets/70325326/a48adeaf-4190-457f-a7e8-1711a4283790)
![Screenshot 2023-05-30 at 23 23 22](https://github.com/orchidsoftware/platform/assets/70325326/61f5c36a-d6e9-4312-962b-908bf667ec19)

i could not make it work.

**Reply by tabuna** [reply]
I am interested in a solution and help.
Can you show me the minimum code to reproduce?

**Reply by macit-emre** [reply]
okay, i will in a short time, thanks in advance.

**Reply by macit-emre** [reply, marked as answer]
here i am again. I have city and district selectboxes. By default district selectbox is empty. My misunderstanding is - as there is no example of how to use Listeners but a simple subtract sample - i thought that layouts method inside Listener wouldn't work anymore. I currently understand this. 
But i see now that, layout method works but a different way. It works as a Layout in normal request like bedfore, but it works from a "Repository based class" when request comes from async. Thus i need to change my Listener's layout that expects Model for logic operations to repository based arrays.

in Listener.php build method there is : $this->query = $repository; 

I now fully understand the Listener working mechanizm. Hoping you not to change anything on Listener again :)

Thanks a lot.

**Comment by gioacchinopoletto** [comment]
I have a similar problem, converting old listeners to new format.
I will receive this error:  

`local.ERROR: Undefined array key 0 {"userId":1,"exception":"[object] (ErrorException(code: 0): Undefined array key 0 at /Users/gpoletto/Sites/thetree/app/Orchid/Layouts/PaymentRequest/PaymentSumListener.php:75)`. 

file `PaymentSumListener.php` at row 75 have the assignment of the request:  

`[$project_id, $project_cost_id] = $request->all();`

@tabuna take note that **same error** occur when I made a new listener, where I want to replicate the example on documentation:  

`[$minuend, $subtrahend] = $request->all();`

Where I'm wrong?
Many thanks in advance.

**Comment by gioacchinopoletto** [comment]
@tabuna I post my listener code for evaluation.
As write on my previous comment the error:   

`local.ERROR: Undefined array key 0 {"userId":1,"exception":"[object] (ErrorException(code: 0): Undefined array key 0 at /Users/gpoletto/Sites/thetree/app/Orchid/Layouts/PaymentRequest/PaymentSumListener.php:42).`   

still **remain also on replicate example listener** on documentation.  
From console:
```
[Error] Failed to load resource: the server responded with a status of 500 (Internal Server Error) (eyJpdiI6ImtiWmVseXI5emRlRWZTUkZJSW5ZMHc9PSIsInZhbHVlIjoiTkZyQ3UrV2pYK0Q1TFpSdVVrK3NVSFJYalpqamNmV3RJN2xXVTZWY1plblRSa01ESnlaMTZBemxVZUY2c2tjekxXUlBteFNmeWk1Qm5GRTdNWWhrR0E9PSIsIm1hYyI6IjlhMWRmYzc1MTkyYWYxYzVlMTU2Mjc0YTg0MGNjNDBhYjEyZTcxOTYyMTdjYjVhNzg3ODEyZjZlYTg0NTBjOGQiLCJ0YWciOiIifQ==, line 0)
[Error] Unhandled Promise Rejection: AxiosError: Request failed with status code 500
```
  

I don't know if the problem is caused by select that have initial  `->empty('')` value (for me required for call the listener).
```
<?php

namespace App\Orchid\Layouts\TransferRequest;

use App\Models\Project;
use Auth;
use Carbon\Carbon;
use Illuminate\Http\Request;
use Orchid\Screen\Fields\DateTimer;
use Orchid\Screen\Fields\Input;
use Orchid\Screen\Fields\Quill;
use Orchid\Screen\Fields\Select;
use Orchid\Screen\Layouts\Listener;
use Orchid\Screen\Repository;
use Orchid\Support\Facades\Layout;

class TransferRequestSumListener extends Listener
{
    /**
     * List of field names for which values will be listened.
     *
     * @var string[]
     */
    protected $targets = [
        'from_project_id',
    ];

    /**
     * What screen method should be called
     * as a source for an asynchronous request.
     *
     * The name of the method must
     * begin with the prefix "async"
     *
     * @var string
     */
    // Pre Orchid 14
    //protected $asyncMethod = 'asyncPayments';

    public function handle(Repository $repository, Request $request): Repository
    {
        [$from_project_id] = $request->all();

        $p = Project::find($from_project_id);

        // outcomes
        $uscite = ($p->projectaccounting->count() > 0) ? $p->projectaccounting->sum('accounting_sum') : 0;

        // incomes
        $entrate_donazioni = ($p->donations->count() > 0) ? $p->donations->where('pivot.accountable', '=', '1')->sum('pivot.sum') : 0;
        $entrate_extra = ($p->projectincomesto->count() > 0) ? $p->projectincomesto->sum('income_sum') : 0;

        $budget_sum = $entrate_donazioni + $entrate_extra - $uscite;

        return $repository
            ->set('from_project_id', $from_project_id)
            ->set('budget_sum', $budget_sum);
    }

    /**
     * @return Layout[]
     */
    protected function layouts(): iterable
    {
        if (Auth::user()->inRole('admin') || Auth::user()->inRole('manager')) {
            $projects_from_to = Project::whereHas('leaders')
                ->where(function ($q) {
                    $q->where('end_date', '>=', date('Y-m-d'))
                        ->orWhereNull('end_date');
                })
                ->orderBy('title', 'asc')
                ->pluck('title', 'id');
        } else {
            $projects_from_to = Project::whereHas('leaders', function ($query) {
                $query->where('leader_id', '=', Auth::user()->id);
            })
                ->where(function ($q) {
                    $q->where('end_date', '>=', date('Y-m-d'))
                        ->orWhereNull('end_date');
                })
                ->orderBy('title', 'asc')
                ->pluck('title', 'id');
        }

        return [

            Layout::rows([

                Select::make('from_project_id')
                    ->title('Project (from)')
                    ->options($projects_from_to)
                    ->empty('')
                    ->required()
                    ->help('Specify the project for transfer request'),

                Input::make('budget_sum')
                    ->title('Project available sum')
                    ->canSee($this->query->has('budget_sum'))
                    ->help('Only for reference: you can ask every sum'),

                Select::make('to_project_id')
                    ->title('Project (to)')
                    ->options($projects_from_to)
                    ->empty('')
                    ->required()
                    ->help('Specify the funds destination project'),

                Input::make('requested_sum')
                    ->type('text')
                    ->title('Sum requested')
                    ->inputmode('decimal')
                    ->required()
                    ->mask([
                        'mask' => '9999999999.99',
                        'numericInput' => true,
                        'digitsOptional' => true,
                        'placeholder' => '',
                    ])
                    ->help(__('Start typing...')),

                DateTimer::make('request_date')
                    ->title(__('Request date extended (only admins and manager)'))
                    ->allowInput()
                    ->format('Y-m-d')
                    ->max(Carbon::now())
                    ->help('This request date have effects on corrisponding accounting row')
                    ->canSee(Auth::user()->inRole('admin') || Auth::user()->inRole('manager')),

                Quill::make('transfer_notes')
                    ->title('Payment transfer notes'),

            ]),

        ];
    }
}

```

**Comment by gioacchinopoletto** [comment]
@tabuna excuse me: **I'm very stupid...** 
With simple _destructure_ function as Laravel macro attached to `Arr` helper we can do the work:  

On /app/Providers:  

```
<?php

namespace App\Providers;

use Illuminate\Support\Arr;
use Illuminate\Support\ServiceProvider;

class AppServiceProvider extends ServiceProvider
{
    /**
     * Register any application services.
     *
     * @return void
     */
    public function register()
    {
        
    }

    /**
     * Bootstrap any application services.
     *
     * @return void
     */
    public function boot()
    {
        Arr::macro(
            'destructure',
            /**
             * Destructure an array from the given keys.
             *
             * @param  array  $array
             * @param  array|string  $keys
             * @return array
             */
            function (array $array, $keys): array {
                $keys = Arr::wrap($keys);
                $results = [];

                foreach ($keys as $key) {
                    if (is_array($key)) {
                        $results[] = Arr::only($array, $key);
                    } else {
                        $results[] = array_key_exists($key, $array) ? $array[$key] : null;
                    }
                }

                $results[] = Arr::except($array, Arr::flatten($keys));

                return $results;
            }
        );
    }
}

```

and on our listener (please see my previous comment):  

`[$from_project_id] = Arr::destructure($request->all(), ['from_project_id']);`

This solve my problem and I hope could be useful also for @macit-emre

**Reply by nuha087** [reply]
this solve my problem too. Thank you so much in advance @gioacchinopoletto
